### PR TITLE
Billable calls entity

### PR DESCRIPTION
**Pull request by mmadariaga**

### CODE REVIEW

File: library/Ivoz/Provider/Domain/Service/BillableCall/MigrateFromTrunksCdr.php [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+use Ivoz\Provider\Domain\Model\BillableCall\BillableCallDto;
+use Ivoz\Provider\Domain\Model\Carrier\CarrierInterface;
+use Ivoz\Provider\Domain\Model\Destination\DestinationInterface;
+use Ivoz\Provider\Domain\Model\RatingPlan\RatingPlanInterface;
+use Ivoz\Provider\Domain\Model\RatingPlan\RatingPlanRepository;
+use Psr\Log\LoggerInterface;
+
+class MigrateFromTrunksCdr
+{
+    const BATCH_SIZE = 5;
```
Comment (Kaian) [inline library/Ivoz/Provider/Domain/Service/BillableCall/MigrateFromTrunksCdr.php]: Increase to 100 at least.

File: library/Ivoz/Provider/Domain/Service/BillableCall/MigrateFromTrunksCdr.php [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+            : '';
+
+        /**
+         * @var TrunksCdrDto $trunksCdrDto
+         */
+        $trunksCdrDto = $this->entityTools->entityToDto(
+            $trunksCdr
+        );
+
+        $caller = $trunksCdrDto->getDiversion()
```
Comment (Kaian) [inline library/Ivoz/Provider/Domain/Service/BillableCall/MigrateFromTrunksCdr.php]: Beware empty getDiversion() values

File: library/Ivoz/Cgr/Infrastructure/Persistence/Doctrine/TpCdrDoctrineRepository.php [hunk 3]
```diff
@@ -36,8 +37,39 @@ public function getOneByCgrid(string $cgrid)
...
+
+    /**
+     * @param string $cgrid
+     * @return int
+     */
+    public function getCarrierRunByCgrid(string $cgrid)
+    {
+        $conditions = [
+            ['cgrid', 'eq', $cgrid],
+            ['runId', 'eq', 'carrier']
```
Comment (cruzccl) [inline library/Ivoz/Cgr/Infrastructure/Persistence/Doctrine/TpCdrDoctrineRepository.php]: Add filter: ['request_type', 'eq', '*postpaid']

File: microservices/scheduler/app/config/services.yml [hunk 4]
```diff
@@ -10,5 +10,7 @@ services:
     autoconfigure: true
     public: true
 
-  Controller:
-    class: "Controller"
+  BillableCallController:
+    class: "BillableCallController"
+  InvoiceSchedulerController:
+      class: "InvoiceSchedulerController"
```
Comment (Kaian) [inline microservices/scheduler/app/config/services.yml]: Extra spaces for indentation

File: scheme/app/DoctrineMigrations/Version20180809101240.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+              INDEX IDX_E6F2DA352480E723 (companyId), 
+              INDEX IDX_E6F2DA356709B1C (carrierId), 
+              INDEX IDX_E6F2DA35BF3434FC (destinationId), 
+              INDEX IDX_E6F2DA355C17F7F9 (ratingPlanId), 
+              INDEX IDX_E6F2DA353D7BDC51 (invoiceId), 
+              INDEX IDX_E6F2DA353B9439A5 (trunksCdrId), 
+              PRIMARY KEY(id)
+          ) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB');
+
+        // Migrating exiting data from kam_trunks_cdrs
```
Comment (Kaian) [inline scheme/app/DoctrineMigrations/Version20180809101240.php]: s/exiting/existing/g

File: scheme/app/DoctrineMigrations/Version20180809101240.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        // Migrating existing data from kam_trunks_cdrs
+        $this->addSql('INSERT INTO BillableCalls (
+                              callid, startTime, duration,
+                              caller, callee, price, priceDetails,
+                              brandId, companyId, carrierId, invoiceId, trunksCdrId
+                          ) SELECT callid, start_time, duration,
+                            COALESCE(NULLIF(diversion,""), caller), callee, price, priceDetails,
+                            brandId, companyId, carrierId, invoiceId, id
+                          FROM kam_trunks_cdrs');
```
Comment (cruzccl) [inline scheme/app/DoctrineMigrations/Version20180809101240.php]: Add WHERE direction='outbound' clause in SELECT

File: scheme/app/DoctrineMigrations/Version20180809101240.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        $this->addSql('ALTER TABLE BillableCalls ADD CONSTRAINT FK_E6F2DA352480E723 FOREIGN KEY (companyId) REFERENCES Companies (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE BillableCalls ADD CONSTRAINT FK_E6F2DA356709B1C FOREIGN KEY (carrierId) REFERENCES Carriers (id) ON DELETE SET NULL');
+        $this->addSql('ALTER TABLE BillableCalls ADD CONSTRAINT FK_E6F2DA35BF3434FC FOREIGN KEY (destinationId) REFERENCES Destinations (id) ON DELETE SET NULL');
+        $this->addSql('ALTER TABLE BillableCalls ADD CONSTRAINT FK_E6F2DA355C17F7F9 FOREIGN KEY (ratingPlanId) REFERENCES RatingPlans (id) ON DELETE SET NULL');
+        $this->addSql('ALTER TABLE BillableCalls ADD CONSTRAINT FK_E6F2DA353D7BDC51 FOREIGN KEY (invoiceId) REFERENCES Invoices (id) ON DELETE SET NULL');
+        $this->addSql('ALTER TABLE BillableCalls ADD CONSTRAINT FK_E6F2DA353B9439A5 FOREIGN KEY (trunksCdrId) REFERENCES kam_trunks_cdrs (id) ON DELETE SET NULL');
+
+        $this->addSql('ALTER TABLE kam_trunks_cdrs ADD metered TINYINT(1) DEFAULT \'0\'');
+        // Mark all CDR entries as metered
+        $this->addSql('UPDATE kam_trunks_cdrs SET metered = 1');
```
Comment (cruzccl) [inline scheme/app/DoctrineMigrations/Version20180809101240.php]: Add WHERE direction='outbound' clause in SELECT

File: debian/systemd/ivozprovider-cdrs.service [hunk 8]
```diff
@@ -0,0 +1,6 @@
+[Unit]
+Description=Run Ivozprovider CDR mixer
+
+[Service]
+Type=oneshot
+ExecStart=/opt/irontec/ivozprovider/microservices/scheduler/bin/run-billbable-calls
```
Comment (cruzccl) [inline debian/systemd/ivozprovider-cdrs.service]: Typo: billable

File: library/Ivoz/Provider/Domain/Service/BillableCall/MigrateFromTrunksCdr.php [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                $trunksCdrDto->getCallid()
+            )->setCaller(
+                $caller
+            )->setCallee(
+                $trunksCdrDto->getCallee()
+            )->setStartTime(
+                $trunksCdrDto->getStartTime()
+            )->setDuration(
+                $trunksCdrDto->getDuration()
+            )->setCallee(
```
Comment (cruzccl) [inline library/Ivoz/Provider/Domain/Service/BillableCall/MigrateFromTrunksCdr.php]: Duplicate

**Pull request merged**